File: Distribute-Coins-in-Binary-Tree/DCIBT.py
class TreeNode:
        def __init__(self, x):
            self.val = x
            self.left = None


class Solution:
    def distributeCoins(self, root: TreeNode) -> int:
        total = 0

        def dfs(node):
            if not node:
                return 0
            L, R = dfs(node.left), dfs(node.right)
            total += abs(L) + abs(R)
            return node.val + L + R - 1

        dfs(root)
        return total

File: Distribute-Coins-in-Binary-Tree/DCIBT.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func distributeCoins(root *TreeNode) int {
	total := 0
	dfs(&total, root)
	return total
}

func dfs(total *int, root *TreeNode) int {
	if root == nil {
		return 0
	}

	L := dfs(total, root.Left)
	R := dfs(total, root.Right)
	if L <= 0 {
		*total += -L
	} else {
		*total += L
	}
	if R <= 0 {
		*total += -R
	} else {
		*total += R
	}

	return root.Val + L + R - 1
}

func main() {}
